File: apps/android/TransformerBridge.kt
package com.transformer.demo

import android.util.Log
import java.nio.ByteBuffer

/**
 * Kotlin bridge to Rust transformer library via JNI
 */
class TransformerBridge {
    
    private var nativeHandle: Long = 0
    
    companion object {
        private const val TAG = "TransformerBridge"
        
        init {
            try {
                System.loadLibrary("transformer_ffi")
                Log.d(TAG, "Loaded native library")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load native library", e)
            }
        }
        
        // Native methods
        @JvmStatic
        external fun nativeInit(): Long
        
        @JvmStatic
        external fun nativeFree(handle: Long)
        
        @JvmStatic
        external fun nativeCreateDefaultModel(handle: Long): Int
        
        @JvmStatic
        external fun nativeLoadModel(handle: Long, data: ByteArray): Int
        
        @JvmStatic
        external fun nativeForward(handle: Long, input: FloatArray): FloatArray?
        
        @JvmStatic
        external fun nativeGetBackend(handle: Long): String
        
        @JvmStatic
        external fun nativeGetVocabSize(handle: Long): Int
    }
    
    /**
     * Initialize the transformer engine
     */
    fun init(): Boolean {
        nativeHandle = nativeInit()
        val success = nativeHandle != 0L
        Log.d(TAG, "Initialization: ${if (success) "success" else "failed"}")
        return success
    }
    
    /**
     * Free resources
     */
    fun free() {
        if (nativeHandle != 0L) {
            nativeFree(nativeHandle)
            nativeHandle = 0
        }
    }
    
    /**
     * Create a default model for testing
     */
    fun createDefaultModel(): Boolean {
        if (nativeHandle == 0L) return false
        val result = nativeCreateDefaultModel(nativeHandle)
        return result == 0
    }
    
    /**
     * Load model from bytes
     */
    fun loadModel(data: ByteArray): Boolean {
        if (nativeHandle == 0L) return false
        val result = nativeLoadModel(nativeHandle, data)
        return result == 0
    }
    
    /**
     * Run inference
     */
    fun forward(input: FloatArray): FloatArray? {
        if (nativeHandle == 0L) return null
        return nativeForward(nativeHandle, input)
    }
    
    /**
     * Get the GPU backend being used
     */
    fun getBackend(): String {
        if (nativeHandle == 0L) return "Unknown"
        return nativeGetBackend(nativeHandle)
    }
    
    /**
     * Get vocabulary size
     */
    fun getVocabSize(): Int {
        if (nativeHandle == 0L) return 0
        return nativeGetVocabSize(nativeHandle)
    }
    
    override fun finalize() {
        free()
    }
}

File: apps/ios/TransformerBridge.swift
import Foundation

/// Swift wrapper for the Rust transformer FFI
class TransformerBridge {
    private var context: OpaquePointer?
    
    init?() {
        context = transformer_init()
        guard context != nil else {
            return nil
        }
    }
    
    deinit {
        if let ctx = context {
            transformer_free(ctx)
        }
    }
    
    /// Create a default model for testing
    func createDefaultModel() -> Bool {
        guard let ctx = context else { return false }
        return transformer_create_default_model(ctx) == 0
    }
    
    /// Load model from bytes
    func loadModel(data: Data) -> Bool {
        guard let ctx = context else { return false }
        return data.withUnsafeBytes { ptr in
            guard let baseAddress = ptr.baseAddress else { return false }
            return transformer_load_model(ctx, baseAddress.assumingMemoryBound(to: UInt8.self), data.count) == 0
        }
    }
    
    /// Run inference
    func forward(input: [Float]) -> [Float]? {
        guard let ctx = context else { return nil }
        
        let vocabSize = Int(transformer_get_vocab_size(ctx))
        guard vocabSize > 0 else { return nil }
        
        var output = [Float](repeating: 0.0, count: vocabSize)
        
        let result = input.withUnsafeBufferPointer { inputPtr in
            output.withUnsafeMutableBufferPointer { outputPtr in
                transformer_forward(ctx, inputPtr.baseAddress, input.count, outputPtr.baseAddress, output.count)
            }
        }
        
        return result == 0 ? output : nil
    }
    
    /// Get the GPU backend being used
    func getBackend() -> String {
        guard let ctx = context else { return "Unknown" }
        guard let cString = transformer_get_backend(ctx) else { return "Unknown" }
        return String(cString: cString)
    }
    
    /// Get vocabulary size
    func getVocabSize() -> Int {
        guard let ctx = context else { return 0 }
        let size = transformer_get_vocab_size(ctx)
        return size > 0 ? Int(size) : 0
    }
}
